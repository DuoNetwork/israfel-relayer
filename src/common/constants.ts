export const AWS_DYNAMO_API_VERSION = '2012-10-08';
export const TAKER_ETH_DEPOSIT = 10; // for development only
export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const PENDING_HOURS = 24;
export const PRUNE_INTERVAL = 3600000;
export const ORDER_PRUNE = 'pruneOrder';
export const SET_ALLOWANCE = 'setAllowance';
export const START_RELAYER = 'startRelayer';
export const PRICE_PRECISION = 8;
export const MIN_ORDER_BOOK_LEVELS = 4;
export const MIN_SIDE_LIQUIDITY = 50;
export const MIN_ETH_BALANCE = 3;
export const MIN_WETH_BALANCE = 2;
export const TARGET_WETH_BALANCE = 6;
export const MAX_WETH_BALANCE = 8;
export const MIN_TOKEN_BALANCE = 100;
export const TARGET_TOKEN_BALANCE = 300;
export const MAX_TOKEN_BALANCE = 400;
export const FAUCET_ADDR = '0x00D8d0660b243452fC2f996A892D3083A903576F';
