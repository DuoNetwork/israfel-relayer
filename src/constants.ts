export const DB_ORDERS = 'orders';
export const DB_MAKER_ADDR = 'makerAddr';
export const DB_TIMESTAMP = 'timestamp';
export const DB_UPDATED_AT = 'updatedAt';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const RELAYER_HTTP_URL = 'http://localhost:3000/v0';
export const RELAYER_WS_URL = 'ws://localhost:8080';
export const MNEMONIC = 'concert load couple harbor equip island argue ramp clarify fence smart topic'
export const BASE_DERIVATION_PATH =  `44'/60'/0'/0`;
export const PENDING_HOURS = 24;

export const WS_CHANNEL_ORDERBOOK = 'orerbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
