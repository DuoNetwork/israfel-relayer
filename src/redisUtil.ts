// import Redis from 'ioredis';
// import * as CST from '../common/constants';
// import { IOrderBook, ITrade } from '../common/types';
// import util from './util';

// export class RedisUtil {
// 	private redisPub: Redis.Redis | null = null;
// 	private redisSub: Redis.Redis | null = null;
// 	private handleOrderBooksUpdate:
// 		| ((channel: string, orderBooks: { [src: string]: IOrderBook }) => any)
// 		| null = null;
// 	private handleTradesUpdate: ((channel: string, trades: ITrade[]) => any) | null = null;

// 	public init(redisKey: { host: string; password: string; servername: string }) {
// 		this.redisPub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
// 		this.redisSub.on('pmessage', (pattern, channel, message) =>
// 			this.onMessage(channel, message, pattern)
// 		);
// 	}

// 	public onMessage(channel: string, message: string, pattern: string = '') {
// 		util.logDebug(pattern + channel);
// 		const [type, ...others] = channel.split('|');
// 		switch (type) {
// 			case CST.DB_ORDER_BOOKS:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, { [others[0]]: JSON.parse(message) });
// 				break;
// 			case CST.DB_OB_OVERVIEW:
// 			case CST.DB_OB_GLOBAL:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, JSON.parse(message));
// 				break;
// 			case CST.DB_TRADES:
// 				if (this.handleTradesUpdate) this.handleTradesUpdate(channel, JSON.parse(message));
// 				break;
// 			default:
// 				break;
// 		}
// 	}

// 	public onOrderBooks(
// 		handleOrderBooksUpdate: (channel: string, orderBooks: { [src: string]: IOrderBook }) => any
// 	) {
// 		this.handleOrderBooksUpdate = handleOrderBooksUpdate;
// 	}

// 	public onTrades(handleTradesUpdate: (channel: string, trades: ITrade[]) => any) {
// 		this.handleTradesUpdate = handleTradesUpdate;
// 	}

// 	public publish(channel: string, msg: string): void {
// 		if (this.redisPub) this.redisPub.publish(channel, msg);
// 	}

// 	public subscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.subscribe(channel);
// 	}

// 	public patternSubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.psubscribe(pattern);
// 	}

// 	public unsubscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.unsubscribe(channel);
// 	}

// 	public patternUnsubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.punsubscribe(pattern);
// 	}

// 	public push(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.lpush(key, values);
// 	}

// 	public putBack(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.rpush(key, values);
// 	}

// 	public async pop(key: string): Promise<string> {
// 		if (this.redisSub) return await this.redisSub.rpop(key);
// 		return '';
// 	}

// 	public async getQueueLength(key: string): Promise<number> {
// 		if (this.redisSub) return this.redisSub.llen(key);
// 		return 0;
// 	}
// }

// const redisUtil = new RedisUtil();
// export default redisUtil;
