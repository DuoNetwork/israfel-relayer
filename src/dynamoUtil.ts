import { SignedOrder } from '0x.js';
import AWS from 'aws-sdk';
import {
	AttributeMap,
	DeleteItemInput,
	PutItemInput,
	QueryInput,
	QueryOutput,
	ScanInput,
	ScanOutput,
	UpdateItemInput
} from 'aws-sdk/clients/dynamodb';
import * as CST from './constants';
import { IDuoSignedOrder, ILiveOrders, UserOrderOperation } from './types';
import util from './util';

class DynamoUtil {
	private ddb: undefined | AWS.DynamoDB = undefined;
	// private process: string = 'UNKNOWN';
	private live: boolean = false;
	private hostname: string = 'hostname';
	private tool: string = 'tool';
	// private contractUtil: ContractUtil | undefined = undefined;
	public init(
		config: object,
		live: boolean,
		tool: string
		//  process: string
	) {
		this.live = live;
		// this.process = process;
		this.tool = tool;
		AWS.config.update(config);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
		return Promise.resolve();
	}

	public  insertData(params: PutItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.putItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public queryData(params: QueryInput): Promise<QueryOutput> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.query(params, (err, data) => (err ? reject(err) : resolve(data)))
					: reject('dynamo db connection is not initialized')
		);
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.scan(params, (err, data) => (err ? reject(err) : resolve(data)))
					: reject('dynamo db connection is not initialized')
		);
	}

	public deleteData(params: DeleteItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.deleteItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public updateItem(params: UpdateItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.updateItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public convertDuoSignedOrderToDynamo(
		order: SignedOrder,
		orderHash: string,
		timestamp: number,
		side: string
	) {
		return {
			[CST.DB_ORDER_HASH]: { S: orderHash },
			[CST.DB_PRICE]: {
				N: order.makerAssetAmount.div(order.takerAssetAmount).valueOf() + ''
			},
			[CST.DB_FILLED_TAKER_ASSET_AMT]: { S: '0' },
			[CST.DB_REMAINING_MAKER_ASSET_AMT]: { S: order.makerAssetAmount.valueOf() + '' },
			[CST.DB_REMAINING_TAKER_ASSET_AMT]: { S: order.takerAssetAmount.valueOf() + '' },
			[CST.DB_SIDE]: { S: side },
			[CST.DB_UPDATED_AT]: { N: timestamp + '' }
		};
	}

	public async addLiveOrder(
		order: SignedOrder,
		orderHash: string,
		marketId: string,
		side: string
	) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS

		const data = this.convertDuoSignedOrderToDynamo(order, orderHash, systemTimestamp, side);

		const params = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_DEV}`,
			Item: {
				[CST.DB_PAIR]: {
					S: marketId
				},
				...data
			}
		};

		await this.insertData(params);
	}

	public async removeLiveOrder(pair: string, orderHash: string): Promise<void> {
		return this.deleteData({
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_DEV}`,
			Key: {
				[CST.DB_PAIR]: {
					S: pair
				},
				[CST.DB_ORDER_HASH]: {
					S: orderHash
				}
			}
		});
	}

	public async deleteOrderSignature(orderHash: string): Promise<void> {
		return this.updateItem({
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.${CST.DB_DEV}`,
			Key: {
				[CST.DB_ORDER_HASH]: {
					S: orderHash
				}
			},
			ExpressionAttributeNames: {
				[CST.DB_SIGNATURE]: CST.DB_SIGNATURE
			},
			ExpressionAttributeValues: {
				[':' + CST.DB_SIGNATURE]: { S: '' }
			},
			UpdateExpression: `SET ${CST.DB_SIGNATURE} = ${':' + CST.DB_SIGNATURE}`
		});
	}

	public convertSignedOrderToDynamo(order: SignedOrder, timestamp: number) {
		return {
			[CST.DB_SENDER_ADDR]: { S: order.senderAddress + '' },
			[CST.DB_MAKER_ADDR]: { S: order.makerAddress + '' },
			[CST.DB_TAKER_ADDR]: { S: order.takerAddress + '' },
			[CST.DB_MAKER_FEE]: { S: order.makerFee.valueOf() + '' },
			[CST.DB_TAKER_FEE]: { S: order.takerFee.valueOf() + '' },
			[CST.DB_MAKER_ASSET_AMT]: { S: order.makerAssetAmount.valueOf() + '' },
			[CST.DB_TAKER_ASSET_AMT]: { S: order.takerAssetAmount.valueOf() + '' },
			[CST.DB_MAKER_ASSET_DATA]: { S: order.makerAssetData + '' },
			[CST.DB_TAKER_ASSET_DATA]: { S: order.takerAssetData + '' },
			[CST.DB_SALT]: { S: order.salt.valueOf() + '' },
			[CST.DB_EXCHANGE_ADDR]: { S: order.exchangeAddress + '' },
			[CST.DB_FEE_RECIPIENT_ADDR]: { S: order.feeRecipientAddress + '' },
			[CST.DB_EXPIRATION_TIME_SECONDS]: { S: order.expirationTimeSeconds.valueOf() + '' },
			[CST.DB_SIGNATURE]: { S: order.signature + '' },
			[CST.DB_UPDATED_AT]: { N: timestamp + '' }
		};
	}

	public async addRawOrder(order: SignedOrder, orderHash: string) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS
		const data = this.convertSignedOrderToDynamo(order, systemTimestamp);

		const params = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.live`
				: `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.dev`,
			Item: {
				[CST.DB_PAIR]: {
					S: orderHash
				},
				...data
			}
		};

		await this.insertData(params);
	}

	public async addUserOrder(
		account: string,
		orderHash: string,
		pair: string,
		operation: UserOrderOperation
	) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS

		const params = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_USER_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_USER_ORDERS}.${CST.DB_DEV}`,
			Item: {
				[CST.DB_ACCOUNT]: {
					S: account
				},
				[CST.DB_PAIR_ORDERHASH]: {
					S: `${pair}|${orderHash}`
				},
				[CST.DB_OPERATION]: {
					S: operation
				},
				[CST.DB_UPDATED_AT]: { N: systemTimestamp + '' }
			}
		};

		await this.insertData(params);
	}

	public parseLiveOrders(data: AttributeMap): ILiveOrders {
		return {
			[CST.DB_PRICE]: Number(data[CST.DB_PRICE].S || '0'),
			[CST.DB_SIDE]: data[CST.DB_SIDE].S || '',
			[CST.DB_AMT]: Number(data[CST.DB_MAKER_ASSET_AMT].S || '0')
		};
	}

	public async getLiveOrders(pair: string): Promise<ILiveOrders[]> {
		const params: QueryInput = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.live`
				: `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.dev`,
			KeyConditionExpression: `${CST.DB_PAIR} = :${CST.DB_PAIR}`,
			ExpressionAttributeValues: {
				[':' + CST.DB_PAIR]: { S: pair }
			}
		};

		const data = await this.queryData(params);
		if (!data.Items || !data.Items.length) return [];

		const parsedLiveOrders = data.Items.map(ob => this.parseLiveOrders(ob)).sort(
			(a, b) => b.price - a.price
		);

		return parsedLiveOrders;
	}

	public updateStatus(process: string) {
		const update: AttributeMap = {
			[CST.DB_STS_PROCESS]: {
				S: `${this.tool}|${process}|${this.hostname}`
			},
			[CST.DB_STS_HOSTNAME]: {
				S: this.hostname
			},
			[CST.DB_UPDATED_AT]: { N: util.getUTCNowTimestamp() + '' }
		};

		return this.insertData({
			TableName: `${CST.DB_PROJECT}.${CST.DB_STATUS}.${this.live ? CST.DB_LIVE : CST.DB_DEV}`,
			Item: update
		}).catch(error => {
			util.logError('Error insert status: ' + error);
			return false;
		});
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
