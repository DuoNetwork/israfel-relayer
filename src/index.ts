import * as CST from './common/constants';
import { IOption } from './common/types';
import orderWatcherServer from './server/orderWatcherServer';
import relayerServer from './server/relayerServer';
import dynamoUtil from './utils/dynamoUtil';
import orderPersistenceUtil from './utils/orderPersistenceUtil';
import osUtil from './utils/osUtil';
import redisUtil from './utils/redisUtil';
import util from './utils/util';
import Web3Util from './utils/Web3Util';

const tool = process.argv[2];
util.logInfo('tool ' + tool);
const option: IOption = util.parseOptions(process.argv);
if (option.debug) util.logLevel = CST.LOG_DEBUG;

const redisConfig = require(`./keys/redis.${option.live ? CST.DB_LIVE : CST.DB_DEV}.json`);
redisUtil.init(redisConfig);

const config = require(`./keys/dynamo.${option.live ? CST.DB_LIVE : CST.DB_DEV}.json`);
dynamoUtil.init(config, option.live, tool, osUtil.getHostName());

let web3Util: Web3Util | null = null;
if ([CST.DB_ORDER_WATCHER, CST.DB_RELAYER].includes(tool))
	web3Util = new Web3Util(null, option.live, '');

switch (tool) {
	case CST.DB_ORDER_WATCHER:
		orderWatcherServer.startOrderWatcher(web3Util as Web3Util, option);
		break;
	case CST.DB_RELAYER:
		relayerServer.startServer(web3Util as Web3Util, option);
		break;
	case CST.DB_ORDERS:
		orderPersistenceUtil.startProcessing(option);
		break;
	default:
		break;
}
