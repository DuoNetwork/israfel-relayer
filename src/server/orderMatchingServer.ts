import { Constants, Util, Web3Util } from '../../../israfel-common/src';
import { IOption } from '../common/types';
import dynamoUtil from '../utils/dynamoUtil';
import orderMatchingUtil from '../utils/orderMatchingUtil';
import redisUtil from '../utils/redisUtil';

class OrderMatchServer {
	public availableAddrs: string[] = [];
	public currentAddrIdx: number = 0;

	public getCurrentAddress() {
		const currentAddr = this.availableAddrs[this.currentAddrIdx];
		this.currentAddrIdx = (this.currentAddrIdx + 1) % this.availableAddrs.length;
		return currentAddr;
	}

	public async startServer(option: IOption) {
		let mnemonic = { mnemomic: '' };
		try {
			mnemonic = require('../keys/mnemomic.json');
		} catch (err) {
			Util.logError(JSON.stringify(err));
		}

		const web3Util = new Web3Util(
			null,
			option.env === Constants.DB_LIVE,
			mnemonic.mnemomic,
			false
		);
		this.availableAddrs = await web3Util.getAvailableAddresses();
		web3Util.setTokens(await dynamoUtil.scanTokens());

		if (option.server) {
			dynamoUtil.updateStatus(
				Constants.DB_ORDER_MATCHER,
				await redisUtil.getQueueLength(orderMatchingUtil.getMatchQueueKey())
			);

			global.setInterval(
				async () =>
					dynamoUtil.updateStatus(
						Constants.DB_ORDER_MATCHER,
						await redisUtil.getQueueLength(orderMatchingUtil.getMatchQueueKey())
					),
				15000
			);
		}

		const loop = async () =>
			orderMatchingUtil.processMatchQueue(web3Util, this.getCurrentAddress()).then(result => {
				global.setTimeout(() => loop(), result ? 0 : 500);
			});
		await loop();
	}
}

const orderMatchServer = new OrderMatchServer();
export default orderMatchServer;
