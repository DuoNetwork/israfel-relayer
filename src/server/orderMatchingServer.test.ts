import * as Constants from '../../../israfel-common/src/constants';
import Util from '../../../israfel-common/src/Util';
import dynamoUtil from '../utils/dynamoUtil';
import orderMatchingUtil from '../utils/orderMatchingUtil';
import redisUtil from '../utils/redisUtil';
import orderMatchingServer from './orderMatchingServer';

jest.mock('../../../israfel-common/src', () => ({
	Constants: Constants,
	Util: Util,
	Web3Util: jest.fn(() => ({
		getAvailableAddresses: jest.fn(() => Promise.resolve(['addr1', 'addr2', 'addr3'])),
		setTokens: jest.fn()
	}))
}));

import { Web3Util } from '../../../israfel-common/src';

test('startProcessing', async () => {
	global.setTimeout = jest.fn();
	dynamoUtil.scanTokens = jest.fn(() =>
		Promise.resolve([
			{
				custodian: 'custodian',
				address: 'address',
				code: 'code',
				denomination: 0.1,
				precisions: {
					WETH: 0.001
				},
				feeSchedules: {
					WETH: {
						minimum: 0,
						rate: 0.01
					}
				}
			}
		])
	);
	dynamoUtil.updateStatus = jest.fn();
	redisUtil.getQueueLength = jest.fn(() => Promise.resolve(100));
	global.setInterval = jest.fn();
	let result = false;
	orderMatchingUtil.processMatchQueue = jest.fn(() => Promise.resolve(result));

	await orderMatchingServer.startServer({ server: true, env: 'live' } as any);
	expect((Web3Util as any).mock.calls).toMatchSnapshot();
	expect((global.setInterval as jest.Mock).mock.calls).toMatchSnapshot();
	await (global.setInterval as jest.Mock).mock.calls[0][0]();
	expect((dynamoUtil.updateStatus as jest.Mock).mock.calls).toMatchSnapshot();
	expect((redisUtil.getQueueLength as jest.Mock).mock.calls).toMatchSnapshot();
	result = true;
	await (global.setTimeout as jest.Mock).mock.calls[0][0]();
	expect((global.setTimeout as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startProcessing, no serveer', async () => {
	dynamoUtil.scanTokens = jest.fn(() =>
		Promise.resolve([
			{
				custodian: 'custodian',
				address: 'address',
				code: 'code',
				denomination: 0.1,
				precisions: {
					WETH: 0.001
				},
				feeSchedules: {
					WETH: {
						minimum: 0,
						rate: 0.01
					}
				}
			}
		])
	);
	dynamoUtil.updateStatus = jest.fn();
	redisUtil.getQueueLength = jest.fn(() => Promise.resolve(100));

	global.setTimeout = jest.fn();
	global.setInterval = jest.fn();
	orderMatchingUtil.processMatchQueue = jest.fn(() => Promise.resolve(false));

	await orderMatchingServer.startServer({ server: false, env: 'dev' } as any);
	expect((Web3Util as any).mock.calls).toMatchSnapshot();
	expect(dynamoUtil.updateStatus as jest.Mock).not.toBeCalled();
	expect(redisUtil.getQueueLength as jest.Mock).not.toBeCalled();
	expect(global.setInterval as jest.Mock).not.toBeCalled();
});
