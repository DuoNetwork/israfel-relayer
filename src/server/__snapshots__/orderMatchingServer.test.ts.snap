// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`startProcessing 1`] = `
Array [
  Array [
    null,
    false,
    "",
    false,
  ],
]
`;

exports[`startProcessing 2`] = `
Array [
  Array [
    [Function],
    15000,
  ],
]
`;

exports[`startProcessing 3`] = `
Array [
  Array [
    "orderMatcher",
    100,
  ],
  Array [
    "orderMatcher",
    100,
  ],
]
`;

exports[`startProcessing 4`] = `
Array [
  Array [
    "match|queue",
  ],
  Array [
    "match|queue",
  ],
]
`;

exports[`startProcessing 5`] = `
Array [
  Array [
    [Function],
    500,
  ],
  Array [
    [Function],
    0,
  ],
]
`;
